Remove duplicated work display from header / profile work

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Bug Fix

## Description

We are currently displaying work information twice:

![before](https://user-images.githubusercontent.com/47985/125416837-ccf023e3-b287-4f5e-b5c4-e4a447ed340c.png)

This PR removes the old display made up of the `employment_title`, `employer_name`, and `employer_url` and only displays the information from the `work` profile field. This means that users who filled in their work information since we deployed the DUS may lose some information, should we do another round of backfilling?

Note: I ended up pulling some changes from https://github.com/forem/forem/pull/14079 into this PR but that's probably for the better since it will reduce the surface area of that cleanup PR. We also deployed a quickfix PR that took care of the issue so I had more time do work on this.

## Related Tickets & Documents

https://github.com/forem/forem/issues/14188

## QA Instructions, Screenshots, Recordings

### QA instructions

1. The "Work" section on `/settings` should look like this (no separate employer fields, one work field):
   ![image](https://user-images.githubusercontent.com/47985/125569170-3ae3538d-069a-41fd-9880-18b9acd23996.png)
2. The work field can be updated
3. The work section on the profile page should look like this (note that the employer name is no longer a link):
    ![image](https://user-images.githubusercontent.com/47985/125570627-24549827-082c-48c6-a6a3-a85be7f4dc3c.png)

### Screenshots

Fixed profile page:

![update profile page](https://user-images.githubusercontent.com/47985/125417326-56c015b5-f014-4d18-88f3-74b47156b39b.png)

Updated profile edit form:

![updated profile edit form](https://user-images.githubusercontent.com/47985/125418654-5101b473-a3ec-4f5d-a94f-6eee5db97151.png)

### UI accessibility concerns?

None

## Added/updated tests?

- [X] Yes, the DUS have tests but I did not add/modify any specs or Cypress tests for the profile page or profile settings page

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

citizen428 commented on app/views/users/_metadata.html.erb:
```diff
@@ -11,21 +11,5 @@
         </p>
       </div>
     <% end %>
-    <% if @user.employment_title.present? %>
```
The `work` field automatically gets picked up by `@profile.ui_attributes_for(area: :header)` so we can remove this special handling.

citizen428 commented on app/views/users/_profile.html.erb:
```diff
@@ -58,7 +58,7 @@
     <div class="crayons-field crayons-field--checkbox">
       <%= f.check_box "profile[display_email_on_profile]",
-                      checked: profile.display_email_on_profile,
+                      checked: @user.setting.display_email_on_profile,
```
I noticed this as a leftover from the user settings work and it got migrated over together with the commit I cherry-picked from another branch.

maestromac commented:
I couldn't reproduce any of the failed view/request specs locally :confused:

citizen428 commented on app/controllers/stories_controller.rb:
```diff
@@ -345,9 +345,7 @@ def set_user_json_ld
       image: Images::Profile.call(@user.profile_image_url, length: 320),
       name: @user.name,
       email: @user.setting.display_email_on_profile ? @user.email : nil,
-      jobTitle: @user.employment_title.presence,
```
This change was slated for https://github.com/forem/forem/pull/14079 but I decided to break that down into smaller pieces, starting with removing some fields.

citizen428 commented on app/controllers/stories_controller.rb:
```diff
@@ -413,25 +411,13 @@ def set_organization_json_ld
     }
   end
 
-  def user_works_for
```
Same as https://github.com/forem/forem/pull/14210/files#r669313889

citizen428 commented on app/controllers/stories_controller.rb:
```diff
@@ -413,25 +411,13 @@ def set_organization_json_ld
       @user.twitter_username.present? ? "https://twitter.com/#{@user.twitter_username}" : nil,
       @user.github_username.present? ? "https://github.com/#{@user.github_username}" : nil,
-      @user.website_url,
+      @user.profile.website_url,
```
Same as https://github.com/forem/forem/pull/14210/files#r669313889

citizen428 commented on app/views/users/_meta.html.erb:
```diff
@@ -1,19 +1,19 @@
 <link rel="canonical" href="<%= user_url(@user) %>" />
-<meta name="description" content="<%= @user.summary %>">
+<meta name="description" content="<%= @user.profile.summary %>">
```
All 3 changes in this file were done for the same reason as https://github.com/forem/forem/pull/14210/files#r669313889

citizen428 commented on app/views/users/show.html.erb:
```diff
@@ -60,10 +60,10 @@
         <div class="profile-header__details">
           <h1 class="crayons-title fw-heavy mb-2"><%= @user.name %></h1>
-          <p class="fs-base m:fs-l color-base-90 mb-4 mx-auto max-w-100 m:max-w-75"><%= @user.summary.presence || ["404 bio not found"].sample %></p>
+          <p class="fs-base m:fs-l color-base-90 mb-4 mx-auto max-w-100 m:max-w-75"><%= @user.profile.summary.presence || ["404 bio not found"].sample %></p>
```
All changes done here are for the same reason as https://github.com/forem/forem/pull/14210/files#r669313889

citizen428 commented on db/seeds.rb:
```diff
@@ -71,9 +71,7 @@
 
     user = User.create!(
       name: name,
-      summary: Faker::Lorem.paragraph_by_chars(number: 199, supplemental: false),
```
With delegation to the profile fields removed we should now set this and `website_url` on the profile directly. The e2e seeds had already been updated by someone else.

citizen428 commented on spec/support/seeds/seeds_e2e.rb:
```diff
@@ -38,17 +38,19 @@
-                        location: "Edinburgh",
-                        education: "University of Life"
-                      })
+  user.profile.update(
```
This is just cosmetic, trying to keep this file similar to `db/seeds.rb`.

citizen428 commented on app/views/shared/_profile_card_content.html.erb:
```diff
@@ -29,25 +29,17 @@
         </li>
       <% end %>
-      <% if actor.employment_title.present? %>
+      <% if actor.profile.respond_to?(:work) && actor.profile.work.present? %>
```
@benhalpern When we decide to consolidate the 3 employment profile fields into the single `work` field you opted for a profile field, not a column on the `profiles` table. This necessitates this special case, which will stop working if an admin e.g. deletes the original field and later recreates it with another label like "Employment" or "Occupation". 

On the profile page this is not a problem, as fields with a `display_area` of `header` get special treatment, independent of their name. This her feels a bit iffy though, so I wonder if we actually should change this to something similar to what we have in `app/views/users/_metadata.html.erb`, i.e. all header fields become part of the preview card.

```ruby
<% if (header_fields = @profile.ui_attributes_for(area: :header)).present? %>
  <div class="profile-header__bottom fs-base">
    <% header_fields.sort.each do |title, value| %>
      <% next if Profile.special_attributes.include?(title) %>
      <div class="crayons-definition">
        <strong class="crayons-definition__title">
          <%= sanitized_sidebar title.titleize %>
        </strong>
        <p class="crayons-definition__value">
          <%= sanitized_sidebar value %>
        </p>
      </div>
    <% end %>
  </div>
<% end %>
```

benhalpern replied:
Yes, I agree with this approach.

My only hesitation here is... We don't currently limit the _total number_ of `display_area=header` fields at the moment, right?

IMO a hard limit of **three** total header fields via custom validation would make sense to solve for a number of potential problems on different views if we can't assume a certain limit.

citizen428 replied:
Yes, that was one of the concerns I had too, we can go with a custom validation to limit this to 3. Thanks!

citizen428 commented on app/models/organization.rb:
```diff
@@ -120,6 +120,12 @@ def destroyable?
     organization_memberships.count == 1 && articles.count.zero? && credits.count.zero?
   end
 
+  # NOTE: We use Organization and User objects interchangeably. Since the former
```
This was part of https://github.com/forem/forem/pull/14079 but I moved it here now.

citizen428 commented on spec/lib/data_update_scripts/add_work_profile_field_spec.rb:
```diff
@@ -4,6 +4,8 @@
 )
 
 describe DataUpdateScripts::AddWorkProfileField do
+  before { ProfileField.destroy_by(label: "Work") }
```
Making sure the field doesn't exist as we seed all profile fields from CSV in `rails_helper.rb`:

```ruby
    # NOTE: @citizen428 needed while we delegate from User to Profile to keep
    # spec changes limited for the time being.
    csv = Rails.root.join("lib/data/dev_profile_fields.csv")
    ProfileFields::ImportFromCsv.call(csv)
    Profile.refresh_attributes!
```

This has been around for a while and we have to see what to do with it long-term.

citizen428 commented on spec/workers/moderator/banish_user_worker_spec.rb:
```diff
@@ -4,7 +4,7 @@
 
   describe "#perform" do
-    let(:user) { create(:user, currently_hacking_on: "text is here") }
+    let(:user) { create(:user) }
```
I completely removed the `currently_hacking_on` attribute here as it has nothing to do with the spec and we don't need it for any assertions.

citizen428 commented on app/views/shared/_profile_card_content.html.erb:
```diff
@@ -57,26 +39,29 @@
-            <%= actor.education %>
-          </div>
-        </li>
+      <% if (header_fields = actor.decorate.ui_attributes_for(area: :header)).present? %>
```
We are now automatically adding all header fields (maximum 3) to the cards. /cc @aitchiss

citizen428 commented on app/models/profile_field.rb:
```diff
@@ -20,7 +23,7 @@ class ProfileField < ApplicationRecord
   validates :display_area, presence: true
   validates :input_type, presence: true
   validates :label, presence: true, uniqueness: { case_sensitive: false }
-  validates :show_in_onboarding, inclusion: { in: [true, false] }
```
This column is non-nullable so the boolean validation can go.

citizen428 replied:
Lately, we've been leaning towards not adding these or even removing them. Right now this value only gets populated via a checkbox in a form in the admin section, which we fully control. I'm not sure how that would return a value other than true or false. 

I have no strong feelings about this, I can add it back in given that it isn't directly related to this PR.

jgaskins replied:
@citizen428 Yeah, it's hard to know which to do, tbh. I could argue that we don't know what all will provide this value in the future (for example, we could add an API endpoint for it) and you could argue that I'm overthinking it/YAGNI and we'd both be right. :sob:

citizen428 replied:
I added it back in for now since it isn't related to this PR anyway. We can maybe make a more comprehensive call in the future and then employ the same strategy across all models with boolean columns.

citizen428 commented on app/views/profile_preview_cards/show.json.jbuilder:
```diff
@@ -1,6 +1,6 @@
 json.extract!(
   @user.profile,
-  :summary, :employment_title, :employer_name, :employer_url, :location, :education
+  :summary, :location, :education, :work
```
@aitchiss `summary`and `location` are ok because they are columns on `profiles` and will therefore always be available on all Forems. The same is not true for `education` and `work`: while we do create them by default (at least for now) admins will be able to remove them. I guess we should do something similar to this, e.g. add all header fields (a maximum of three):

```ruby
      <% if (header_fields = actor.profile.decorate.ui_attributes_for(area: :header)).present? %>
        <% header_fields.sort.each do |title, value| %>
          <% next if Profile.special_attributes.include?(title) %>
          <li>
            <div class="key">
              <%= title %>
            </div>
            <div class="value">
              <%= value %>
            </div>
          </li>
        <% end %>
      <% end %>
```

However, this PR is getting big enough as is, so I'll do that later.

citizen428 commented on lib/data_update_scripts/20210630041322_migrate_data_to_work_field.rb:
```diff
@@ -1,19 +0,0 @@
-module DataUpdateScripts
```
Given our recent discussions on forem.team, maybe this shouldn't be removed at all? I'm happy to restore it.

citizen428 commented on spec/models/profile_field_spec.rb:
```diff
@@ -20,5 +20,17 @@
+
+    describe "#maximum_header_field_count" do
+      it "limits the number of header fields" do
+        count = [0, described_class::HEADER_FIELD_LIMIT - described_class.header.count].max
```
While this isn't great to read it made the spec less error-prone as now it doesn't matter how many header fields already exist when running this.

citizen428 commented on spec/system/user/user_edits_profile_spec.rb:
```diff
@@ -15,12 +15,6 @@
            label: "Hate Ice Cream Flavor",
            display_area: "header")
   end
-  let!(:settings_only_profile_field) do
```
I completely removed this because we previously had 3 distinct display areas and this spec tested 1 field in each area. Now that we only have 2 areas, 2 fields are enough.

citizen428 commented:
This is almost ready to go. I still have 2 spec failures on Travis:

1. `spec/models/user_spec.rb:831`: this looks like the delegation from `User` to `Profile` isn't happening. The spec is green locally though and modifying the spec to ensure that the profile field 100% exists didn't help either.
2. The `Should show author details in the preview card` in `articleFlows/previewProfile.spec.js`: on Travis it seems like it can't find the work info, locally it fails on the location, neither of that should happen.

citizen428 commented:
The spec failure on Travis is getting bizarre: during the first run `user` does indeed not respond to `available_for`. 

![image](https://user-images.githubusercontent.com/47985/127279370-b618af20-fa18-486d-b692-9dce8b7d9bdc.png)

What's weird is that the method returns `nil` though because no code path in the C implementation of `respond_to?` allows for that, undefined values explicitly get converted to `false`.

```ruby
if (ret == Qundef) ret = Qfalse;
return ret;
```

djuber commented:
I did a little digging into the spec failure, and am able to recreate it locally by resetting the database prior to the test

the travis config has this
```yaml
script:
  - bundle exec rails db:create db:schema:load webpacker:compile
  - bin/knapsack_pro_rspec
```

A equivalent local reproduction looks like this

```
$ RAILS_ENV=test bundle exec rails db:schema:load
$ bundle exec rspec spec/models/user.rb:833
Run options: include {:locations=>{"./spec/models/user_spec.rb"=>[833]}}
[Zonebie] Setting timezone: ZONEBIE_TZ="International Date Line West"
F

Failures:

  1) User profiles propagates changes of unmapped attributes to the profile model
     Failure/Error: user.update(available_for: "profile migrations")
     
     ActiveModel::UnknownAttributeError:
       unknown attribute 'available_for' for User.
     # ./spec/models/user_spec.rb:838:in `block (4 levels) in <top (required)>'   
     # ./spec/models/user_spec.rb:837:in `block (3 levels) in <top (required)>'   
     # ./spec/rails_helper.rb:147:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:147:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:99:in `block (2 levels) in <top (required)>'

Finished in 0.64305 seconds (files took 2.57 seconds to load)
1 example, 1 failure

Failed examples:
rspec ./spec/models/user_spec.rb:833 # User profiles propagates changes of unmapped attributes to the profile model
```

This sequence of events (clear the db, then run the test) consistently fails for me locally.

The reason, from what I can see, is that we're relying on specific ProfileFields to have been loaded from a csv file (`before(:suite)` block in the rails helper), but we're checking for profile fields in the database (there are none after a schema load, or during the first test run in travis) during User class load.

I added a small amount of debug logging to the loading - it looks like the models are being loaded during `Rails.application.initialize!` in config/environment.rb (which is required by rails helper very early).

I was able to cheat a little and make this test pass by calling `load(Rails.root.join("app/models/user.rb"))` inside the spec (rather than calling `Profile.refresh_attributes!) - this caused the concerning block in user (`included do`) to be re-run, after the seeded profile fields are present because now we're inside the suite, and Profile has the expected attributes. We query profile_fields once when loading the Profile class, then again, when we call Profile.refresh_attributes! when we load user and call the `concerning ProfileMigrations`, and then we start the tests (including the before(:suite) block which creates the profile fields we want to delegate.


Some judicious logging statements show this sequence (I couldn't use the rails logger before rails had been loaded, so there's an open-ended beginnging)

```
$ > log/test.log                                                      

$ bundle exec rspec spec/models/user_spec.rb:833
## pass or fail - it does not matter for this tracing

$ grep ^2021- log/test.log -A 1         
2021-07-28T10:02:47-05:00: User: loading class file
2021-07-28T10:02:47-05:00: Profile: loading class file
2021-07-28T10:02:47-05:00: ProfileField: loading class file
2021-07-28T10:02:47-05:00: ProfileField: loaded class file
  ProfileField Load (0.4ms)  SELECT "profile_fields".* FROM "profile_fields" ORDER BY "profile_fields"."id" ASC LIMIT $1  [["LIMIT", 1000]]
2021-07-28T10:02:47-05:00: Profile: loaded class file
  ProfileField Load (0.2ms)  SELECT "profile_fields".* FROM "profile_fields" ORDER BY "profile_fields"."id" ASC LIMIT $1  [["LIMIT", 1000]]
2021-07-28T10:02:47-05:00: User: delegate profile column available_for
2021-07-28T10:02:47-05:00: User: delegate profile column currently_hacking_on
2021-07-28T10:02:47-05:00: User: delegate profile column currently_learning
2021-07-28T10:02:47-05:00: User: delegate profile column education
2021-07-28T10:02:47-05:00: User: delegate profile column location
2021-07-28T10:02:47-05:00: User: delegate profile column summary
2021-07-28T10:02:47-05:00: User: delegate profile column website_url
2021-07-28T10:02:47-05:00: User: loaded class file
2021-07-28T10:02:47-05:00: RailsHelper: required config/environment and initialized rails
```

The reason the test passes on the second execution, or manually trying to recreate the steps in a console, is that the class was loaded by the process, and the csv file had been loaded by the prior execution (we don't clear the db after testing, which is normally not a problem, and the csv seeder is designed to skip existing profile fields so there's no conflict there). 

Clearing the db (rails db:schema:load) as we do prior to the CI suite running, before calling rspec or starting a console (loading the app in either case) is the most reliable way to reproduce this.

citizen428 commented:
> The reason, from what I can see, is that we're relying on specific ProfileFields to have been loaded from a csv file (before(:suite) block in the rails helper), but we're checking for profile fields in the database (there are none after a schema load, or during the first test run in travis) during User class load.

@djuber Thanks! I'm still not sure why that would start happening now though, the code/test setup has been like this for a couple of months already. I don't want to spend too much time on a solution to this problem because the whole delegation from `User` to `Profile` will go away relatively soon but it bugs me to not understand what in this PR caused this difference in behavior.

citizen428 commented:
So I decided to remove the new guard clause I added in this PR:

```ruby
next unless column.in?(Profile.attributes) || column.in?(Profile::STATIC_FIELDS)
```

We didn't have it before and it's not strictly speaking necessary as I already removed problematic calls. And given that delegation will go away relatively soon I'd rather just finally get this PR done, I already wasted too much time as-is.

Thanks @djuber for making me realize which line caused the problem which allowed me to make a call on how to best resolve this issue.

rhymes commented on app/models/user.rb:
```diff
@@ -636,8 +636,12 @@ def bust_cache
     Users::BustCacheWorker.perform_async(id)
   end
 
+  # TODO: @citizen428 I don't want to completely remove this method yet, as we
+  # have similar methods in other models. But the previous implementation used
+  # three profile fields that we can't guarantee to exist across all Forems. So
+  # for now this method will just return an empty string.
```
I think we can completely remove this method eventually and probably `search_score` as well. Unrelated to this PR though

citizen428 replied:
Yes, cleaning those up is on my todo list.

rhymes reviewed (commented):
The code seems to be okay. I ran the scripts, I can update the fields but only Education shows up.

See the video:


https://user-images.githubusercontent.com/146201/127449753-dab4d6f4-d2b8-4859-bf62-9267514c15a3.mov

I also tried to restart the server, just in case :D 

I loaded my profile in the DB:

```ruby
[3] pry(main)> user.profile
  Profile Load (5.9ms)  SELECT "profiles".* FROM "profiles" WHERE "profiles"."user_id" = $1 LIMIT $2  [["user_id", 13], ["LIMIT", 1]]
=> #<Profile:0x00007fbe421afb50
 id: 13,
 created_at: Thu, 22 Jul 2021 13:22:11.122852000 UTC +00:00,
 data: {"work"=>"Dev at Forem 2", "education"=>"Super high in the sky", "available_for"=>"", "skills_languages"=>"", "currently_learning"=>"", "currently_hacking_on"=>""},
 location: "",
 summary: "",
 updated_at: Thu, 29 Jul 2021 07:25:15.449097000 UTC +00:00,
 user_id: 13,
 website_url: "">
[4] pry(main)> user.profile.data
=> {"work"=>"Dev at Forem 2", "education"=>"Super high in the sky", "available_for"=>"", "skills_languages"=>"", "currently_learning"=>"", "currently_hacking_on"=>""}
[5] pry(main)> user.profile.work
=> "Dev at Forem 2"
```

everything seems to be in the right place

citizen428 commented:
Oops, sorry @rhymes This line obviously needs to be removed now, that's a leftover from the quickfix.

rhymes reviewed (approved):
LGTM!

mstruve reviewed (approved):
Pulled it down and messed around with work and education fields and everything worked great! 🚀

citizen428 commented on app/models/profile_field.rb:
```diff
@@ -35,4 +38,10 @@ def type
   end
+
+  def maximum_header_field_count
+    return if self.class.header.count < HEADER_FIELD_LIMIT
```
We want a maximum of 3 profile fields. So we return here as long as well still have less than that. If we changed the `<` to `<=` then we'd return without adding an error when 3 fields already exist and add a 4th one.

jgaskins replied:
Ahh, okay. No, you understood correctly, but I feel like that order of operations isn't super clear.

Also, is there ever a reason to update them in place? Because this protects against creation after 2, which it seems is where you were coming from, but we can't update any of them when we're currently at the max:

```ruby
[13] pry(main)> ProfileField.header.count
=> 3
[14] pry(main)> ProfileField.last.validate!
ActiveRecord::RecordInvalid: Validation failed: Display area maximum number of header fields (3) exceeded
from /Users/jamie/.rvm/gems/ruby-2.7.2/gems/activerecord-6.1.4/lib/active_record/validations.rb:80:in `raise_validation_error'
```

citizen428 replied:
> Ahh, okay. No, you understood correctly, but I feel like that order of operations isn't super clear.

Agreed, but you hit on that in your other comment, Rubocop kinda forced me into a guard clause which really doesn't work too well here. 

Good point regarding the validation, while we don't generally mutate them in place it is possible via the admin UI, so there's a subtle bug lurking here. I'll set this PR back to draft and update the validation tomorrow, including a test based on the example above. Thanks!

jgaskins replied:
True, this was definitely part of that same confusion.

citizen428 commented:
I updated the validation and added some tests around it. Once again Rubocop sort of forced me into writing this in a very specific way that I wouldn't necessarily have chosen myself.
